Document: Add SMBIOS additional Information for LoongArch (#34)
loongson/LoongArch-Documentation: docs/LoongArch-SMBIOS-Additional-Information-EN.adoc

= LoongArch Processor Additional Information
Loongson Technology Corporation Limited
v1.00
:docinfodir: ../themes
:docinfo: shared
:doctype: book
:toc: left

=== Vendor Name register and CPU Name register

> The LoongArch CPU designed by Loongson has two registers representing Machine Vendor Name and CPU Name, both of which are ASCII string values in DQWORD-formatted, respectively located at offsets 0x10 and 0x20 of the IOCSR register.

Reviewer: DQWORD 是啥，“双四字”，我印象中不是要么 DWORD 要么 QWORD 吗？

另外，“IOCSR”的 R 就是 register 的意思，不需要冗余写“IOCSR register”。这里实际要表达的意思是“IOCSR 地址空间”，而不是单个寄存器（单个寄存器哪来 0x20 这么宽），所以应该写“IOCSR space”。

Author: 好的，这个确实表达不严谨

=== LoongArch Processor-specific Block Structure
[%header,cols="^1,^1,^1,^1,3"]
|===
|Offset
|Name
|Length
|Value
^|Description

|00h
|Block Length
|BYTE
|Varies (N)

> |Length of Processor-specific Data

Reviewer: 这边不用 title-case 吧，首字母大写就行了

|01h
|Processor Type
|BYTE
|Varies
|The processor architecture delineated by this Processor-specific Block. (See SMBIOS Table 131). 0x9 means LoongArch32 and 0x10 means LoongArch64.

|02h
|Processor-Specific Data
|N BYTEs
|Varies
|Processor-specific data.

|03h
|Reserved
|BYTE
|Varies
|Reserved.

|04h
|Machine Vendor ID
|DQWORD
|Varies

> |Vendor ID of this LoongArch processor manufacturer.

Reviewer: 句子有点问题，感觉是不是直接写“Vendor ID of the processor”就够了？

Author: 修改了

|14h
|CPU ID
|DQWORD
|Varies

> |The CPU ID used for this LoongArch processor manufacturer to mark different CPU types or CPU instances.

Reviewer: 名字不考虑起个“Processor ID”跟 LA 手册里提到的 CPUCFG PRID 域命名保持一致吗？

另外后边的解释，感觉既多余又矛盾：多余是指的如果这个字段其实就是冗余一份 CPUCFG PRID，那不需要在这里重复讲解一遍语义，矛盾是指 PRID 字段目前在真实硬件（如 3A5000）上没有设计为产品唯一 ID，软件上也没有这么认为，所以实际上该字段的语义根本不可能实现为文中“or”的语义而只能取其一。

Author: PRID那个域是架构的，跟CPU不是一个维度的，从英语的Processor和CPU的意义来看，我觉得Processor应该是CPU同级别的

|24h
|Extended instruction set support
|DWORD
|Bit-field
|The bit field [3:0] encodes the existing LoongArch standard extensions instruction set. It equivalent to bits [3:0] in LoongArch EUEN register(CSR 0x2). +

> Any bit of this field is set to 1, indicating that the corresponding extension is supported, and 0 indicates not supporting.

Reviewer: 上面已经提了和 EUEn 字段语义一致了，这句解释就可以去掉了。

Author: EUEN的语义是启用与否，这个是支持与否

> = LoongArch Processor Additional Information

Reviewer: Change to be more like the [riscv-smbios](https://github.com/riscv/riscv-smbios) project:

```suggestion
= LoongArch Processor SMBIOS Spec
```

Loongson Technology Corporation Limited
v1.00
:docinfodir: ../themes
:docinfo: shared
:doctype: book
:toc: left

> This document is used to describe additional information about the LoongArch processor of SMBIOS Type 44(Type 44 added after SMBIOS v3.3.0).

Reviewer: 这个句子非常中式英语，帮你改了下：

```suggestion
This document defines the LoongArch processor-specific data block to supplement the upstream definition of SMBIOS structure Type 44 (Processor Additional Information, Section 7.45 of SMBIOS 3.5.0 specification).
```

=== Vendor Name register and CPU Name register

> The LoongArch CPU designed by Loongson has two registers representing Machine Vendor Name and CPU Name, both of which are ASCII string values in DQWORD-formatted(SMBIOS uses 16 bits as a WORD, total 128 bits), respectively located at offsets 0x10 and 0x20 of the IOCSR space.

Reviewer: `The LoongArch CPU designed by Loongson has two registers representing Machine Vendor Name and CPU Name`，那非龙芯公司设计的 LoongArch CPU 行为如何？

另外这一行的括号前面要加一个空格

Author: 这个是架构外的，使用LoongArch指令集完成的实例我们不保证他们会完成IOCSR或者IOCSR和龙芯设计的一样

Reviewer: > 这个是架构外的，使用LoongArch指令集完成的实例我们不保证他们会完成IOCSR或者IOCSR和龙芯设计的一样

但这份文档应该对那些采用非龙芯公司 LoongArch CPU 设计的板卡也适用，可能还是要微调一下措辞，明确无论具体 CPU 对这些寄存器的实现如何，此处假定参照的是龙芯设计 CPU 的行为。

Author: 好的，我一会儿修改一下您再帮忙review一下

Author: 其实这个措辞我想了想还是不用写，因为在下面的表里面已经提到了，在龙芯设计的CPU等价于这两个寄存器，否则表格提到的两个域就是来给CPU manufacturer来用的，这小节就是介绍了龙芯设计的CPU里面有两个寄存器来表征相关信息的。

Reviewer: 那我感觉实际上不用特意强调和 LA 这俩寄存器的联系。。龙芯自己怎么设计 CPU 和自己的板卡，是龙芯的实现细节，标准不用去管（这边的 manufacturer 和 model 字串要不要跟 CPU 里存在的串保持一致是板卡厂商自己的事）

这种情况建议删掉绝大多数文字，包括这一节（在 SMBIOS 规范补充文档里出现“寄存器”的章节很奇怪，仿佛是 SMBIOS 规范要修补 LoongArch 架构文档一样），至多加入“如果 CPU 实现了 IOCSR 0x10/0x20 则建议此处取值与 CPU IOCSR 读出值保持一致”就好了，非常精炼。

> The LoongArch CPU designed by Loongson has two registers representing Machine Vendor Name and CPU Name, both of which are ASCII string values in DQWORD-formatted(SMBIOS uses 16 bits as a WORD, total 128 bits), respectively located at offsets 0x10 and 0x20 of the IOCSR speace.

Reviewer 2: 这行最后一个单词是space?

Author: 眼尖，我改改

=== LoongArch Processor-specific Block Structure
[%header,cols="^1,^1,^1,^1,3"]
|===
|Offset
|Name
|Length
|Value
^|Description

|00h
|Block Length
|BYTE
|Varies (N)
|Length of Processor-specific Data

|01h
|Processor Type
|BYTE
|Varies
|The processor architecture delineated by this Processor-specific Block. (See SMBIOS Table 131). 0x9 means LoongArch32 and 0x10 means LoongArch64.

|02h
|Processor-Specific Data
|N BYTEs
|Varies
|Processor-specific data.

|03h
|Reserved
|BYTE
|Varies
|Reserved.

|04h
|Machine Vendor ID

> |DQWORD

Reviewer: 我查证了 [SMBIOS 规范文档](https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.5.0.pdf)，确认不存在 `DQWORD` 的写法，应该用 `QWORD` 表示“64 位”。请修改一下。

Author: 您已经查到RV的规范了，不妨看看他们的写法，我也是仿照他们的写的DQWORD :)
https://github.com/riscv/riscv-smbios/blob/master/riscv-smbios.pdf

Reviewer: 由于上游规范文档里没提，RV 规范跟我们这篇是平级的，不妨单独弄一小节解释 DQWORD 这东西？我反正真的找了大半天 🤦

Author: 要不然就不提DQWORD，就写two QWORD也可以，上游有这么写的

Reviewer: 那怎么清晰怎么来呗，我的建议是尽量往上游而不是 RV 靠（因为 RV 文档跟我们是平等的关系），但 DQWORD 在解释清楚的前提下更精炼也是事实。

|Varies
|The manufacturer vendor ID of the processor. +

> It is equivalence to IOCSR space offset 0x10 on a Loongson CPU.

Reviewer: ```suggestion
Equal to the value located at IOCSR space offset 0x10.
```

这边把“在龙芯 CPU”几个字先去了，需要根据前面提到的在非龙芯设计的 LoongArch 处理器上行为如何的结论来更新这边的措辞。

Author: 这个还是根据上面提到的，如果修改措辞后还是要写"on a Loongson CPU"，类似于"on a AMD CPU"一个意思

Reviewer: 了解，当然考虑到“龙芯”一词承载的太多意义，这里改成`on a Loongson-designed CPU` 或者 `on a CPU designed by Loongson` 这样好像也差不多，你们自己斟酌一下

Author: 我感觉这里的Loongson就是特指Loongson设计的CPU，因为现在LoongArch和Loongson已经分开了，没什么歧义

|14h
|CPU ID
|DQWORD
|Varies
|The CPU ID used for this LoongArch processor manufacturer to mark different CPU types or CPU instances. +

> It is equivalence to IOCSR space offset 0x20 on a Loongson CPU.

Reviewer: 同上

|24h

> |Extended instruction set support

Reviewer: ```suggestion
|ISA extensions support
```

|DWORD
|Bit-field

> |The bit field [3:0] encodes the existing LoongArch standard extensions instruction set. It equivalent to bits [3:0] in LoongArch EUEN register(CSR 0x2). +

Reviewer: ```suggestion
|The bit field [3:0] indicate support for the existing LoongArch ISA extensions. It is modeled after the LoongArch EUEN register (CSR 0x2), and meanings of individual bits is the same as defined for the EUEN register. +
```

> Any bit of this field is set to 1, indicating that the corresponding extension is supported, and 0 indicates not supporting.

Reviewer: ```suggestion
The corresponding bit is set for every ISA extension supported on this system.
```
